Fix https_batch deadlock due to golang timer changes

nicklas-dohrn opened the pull request:
### What is changed:

This changes the way the syslog batches are triggered.
The new implementation no longer uses Timers that need to be reset and checked, it just ticks once every second per http_batch drain to send a batch at least once a second. 
The new logic is as follows:
- If there is a log present in the queue, send it at last one second later with every log from the same second.
- If there are more logs in one second than the 'batchSize' allows, send the batch immediately 

### The problem fixed by this change:

The old implementation was able to deadlock itself due to the way the time channel was reset. There are even more changes to this behaviour to be expected by the golang 1.23 update. (https://tip.golang.org/doc/go1.23#timer-changes)
-> Changed logic necessary to prevent further issues on further golang upgrades

### Impact:
The new implementation will tick more often, but the overhead will be pretty minimal, due to not doing anything for empty drains.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Testing performed?

- [x] Unit tests
- [x] Tested in dev cluster

## Checklist:

- [x] This PR is being made against the `main` branch, or relevant version branch
- [x] I have made corresponding changes to the documentation
- [x] I have added testing for my changes

juergen-walter commented:
@ctlong can you check please

nicklas-dohrn commented on src/pkg/egress/syslog/https_batch.go:
```diff
@@ -65,34 +65,37 @@ func (w *HTTPSBatchWriter) Write(env *loggregator_v2.Envelope) error {
+
+	sendBatch := func() {
+		if msgBatch.Len() > 0 {
+			w.sendHttpRequest(msgBatch.Bytes(), msgCount)
```
Fixed the issue.
This was done with a //nolint before as well,
so it should be fine to keep it for now.

ctlong reviewed (commented):
Can you write a failing test for the old logic that works for the new logic?

nicklas-dohrn commented:
I am currently in the process to rewrite the tests anyway, but I will make sure to have a test that proves the point, that the old implementation deadlocks/hangs in certain scenarios.

nicklas-dohrn commented:
This change now contains a test case, that will not work with the old timer based implementation.
I added a test configurable defaultTime for the ticker, to keep test cases shorter than I would need with 1 second intervalls.

chombium reviewed (commented):
The code looks good, but we'll have to do some integration tests as well before approving the PR.

nicklas-dohrn commented:
Unfortunately, I cannot see the failures due to the concourse build check.
It would be nice to know, what I should fix here.

juergen-walter commented:
Hi @ameowlia @ctlong not having access nor any information about failing **concourse-ci/tests** is discouraging and a huge delay factor for open source contributors such as @nicklas-dohrn. Any ideas on how to improve? Next PR is already in the pipeline https://github.com/cloudfoundry/loggregator-agent-release/pull/617

ctlong commented:
> Hi @ameowlia @ctlong not having access nor any information about failing **concourse-ci/tests** is discouraging and a huge delay factor for open source contributors such as @nicklas-dohrn. Any ideas on how to improve? Next PR is already in the pipeline #617

Sorry that is blocking progress here. I'm aware that the lack of visibility with these PR tests are not ideal, but there's no quick fixes to expose this pipeline or move it to an open source concourse at the moment.

I'm happy to discuss in Slack the path I'm attempting to take to open source the pipelines. I'll take any help that I can get to implement the plan 😄

ctlong commented:
Here's the failure:

```bash
$ go run github.com/onsi/ginkgo/v2/ginkgo -r --randomize-all --randomize-suites --fail-on-pending --keep-going --race --trace
...
------------------------------
• [FAILED] [0.146 seconds]
HTTPS_batch [It] test dispatching for batches before timewindow is finished
/tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:100

  Timeline >>
  2025/02/18 01:05:36 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/02/18 01:05:36 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/02/18 01:05:36 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/02/18 01:05:36 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  [FAILED] in [It] - /tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:105 @ 02/18/25 01:05:36.995
  << Timeline

  [FAILED] Expected
      <int>: 257
  to equal
      <int>: 256
  In [It] at: /tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:105 @ 02/18/25 01:05:36.995

  Full Stack Trace
    code.cloudfoundry.org/loggregator-agent-release/src/pkg/egress/syslog_test.init.func3.4()
    	/tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:105 +0x4fe
```

Concourse container is an x86_64 Linux OS with 16 cores.

chombium commented:
I've run the test now few times on my machine with Ubuntu 22.04 and from four runs I've got three errors and one successful run.. It seems to me that as the code plays with time ranges and the randomness that we run the test suite with, makes problems. The tests should be more robust, I guess... @nicklas-dohrn wdyt?

chombium commented:
> I'm happy to discuss in Slack the path I'm attempting to take to open source the pipelines. I'll take any help that I can get to implement the plan 😄
@ctlong I would gladly help you with this

ctlong commented:
Yes, it seems like the "test dispatching for batches before timewindow is finished" test assumes that the following lines can be executed within one second:
```Go
env1 := buildLogEnvelope("APP", "1", "string to get log to 1024 characters:"+string_to_1024_chars, loggregator_v2.Log_OUT)
for i := 0; i < 300; i++ {
    Expect(writer.Write(env1)).To(Succeed())
}
Expect(drain.getMessagesSize()).Should(Equal(256))
```

If they take longer than one second, then more envelopes may be present in the drain than just 256.

Why not make the `DefaultSendInterval` very large just for this test and remove the sleep and final expectation?

nicklas-dohrn commented:
I took your inspiration for the testcase:
"test dispatching for batches before timewindow is finished"

but did it kind of the other way around:
if reducing the overall size of the batch, the timings should be less tight, so the fast execution time should work.
I also changed the bigger sized logs to not hit the buffer size limit directly, to account for inconsistencies in object to string materialisation (not sure if this can happen, but now it is also really clear, when the buffer limit is breached)

I hope that with the other changes I introduced will make this fail hopefully never again.

chombium commented:
This looks good. I'll do few more unit test runs tomorrow, so that we are sure that the problem is fixed.

chombium commented on src/pkg/egress/syslog/https_batch.go:
```diff
@@ -11,7 +11,8 @@ import (
 )
 
-const BATCHSIZE = 256 * 1024
+var DefaultBatchSize = 256 * 1024
```
Could you please make these two as constants?

nicklas-dohrn replied:
They cannot be constants, as they need to be set in the Tests.
There seems to be no easy way around doing it like this, so can we not keep it this way?

chombium replied:
For cases like this we're using the so called `Functional options pattern`. An object is created with default values and then options are applied.Sorry for not mentioning it earlier...
Here is an [example definition](https://github.com/cloudfoundry/loggregator-agent-release/blob/main/src/pkg/clientpool/v2/balancer.go#L10-L24) and [usage](https://github.com/cloudfoundry/loggregator-agent-release/blob/main/src/cmd/loggregator-agent/app/app_v2.go#L173] from this project.

You can find more info about this pattern in the following blog posts:
https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c
https://golang.cafe/blog/golang-functional-options-pattern.html

This way, you will be able to have the default values configured and can be overwritten if needed.

chombium commented on src/pkg/egress/syslog/https_batch_test.go:
```diff
@@ -17,7 +17,18 @@ import (
 )
 
-var string_to_1024_chars = "saljdflajsdssdfsdfljkfkajafjajlköflkjöjaklgljksdjlakljkflkjweljklkwjejlkfekljwlkjefjklwjklsdajkljklwerlkaskldgjksakjekjwrjkljasdjkgfkljwejklrkjlklasdkjlsadjlfjlkadfljkajklsdfjklslkdfjkllkjasdjkflsdlakfjklasldfkjlasdjfkjlsadlfjklaljsafjlslkjawjklerkjljklasjkdfjklwerjljalsdjkflwerjlkwejlkarjklalkklfsdjlfhkjsdfkhsewhkjjasdjfkhwkejrkjahjefkhkasdjhfkashfkjwehfkksadfjaskfkhj ...
+var string_to_256_chars string
```
Please rename this variable to be in camel case which is Go's standard naming convention?

nicklas-dohrn replied:
done

chombium reviewed (changes requested):
Hi @nicklas-dohrn, thank you very much for the quick reponse.

I have few more small objections and than I will test the changes once again. Could you please also squash your commits?

nicklas-dohrn commented:
@ctlong if this one runs through, you can merge it now.
should be stable now.

chombium reviewed (changes requested):
Hi @nicklas-dohrn, I have a suggestion how to restructure the code for setting he default values and I was able to make the test break again.

I've run the Ginkgo test 10 times one after another:
```
 for i in {1..10}; do go run github.com/onsi/ginkgo/v2/ginkgo -r --randomize-all --randomize-suites --fail-on-pending --keep-going --race --trace >> /tmp/nicklas.log; done
```
The broken tests are:
```
[1740406427] Syslog Suite - 84/84 specs ••••••••••••
------------------------------
• [FAILED] [0.165 seconds]
HTTPS_batch [It] testing simple appending of one log
/tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:62

  Timeline >>
  2025/02/24 15:13:56 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/02/24 15:13:56 Dropped 10000 dropping logs for test-source-id's app drain with url dropping://my-drain:8080/path
  [FAILED] in [It] - /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:69 @ 02/24/25 15:13:56.394
  << Timeline

  [FAILED] Expected
      <int>: 0
  to equal
      <int>: 2
  In [It] at: /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:69 @ 02/24/25 15:13:56.394

  Full Stack Trace
    code.cloudfoundry.org/loggregator-agent-release/src/pkg/egress/syslog_test.init.func3.2()
    	/tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:69 +0x87e
------------------------------
••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••
------------------------------
• [FAILED] [0.193 seconds]
HTTPS_batch [It] test batch dispatching with all logs in a given timeframe
/tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:96

  Timeline >>
  2025/02/24 15:13:59 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/02/24 15:13:59 Dropped 10000 dropping logs for test-source-id's app drain with url dropping://my-drain:8080/path
  2025/02/24 15:13:59 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/02/24 15:13:59 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/02/24 15:13:59 Dropped 10000 dropping logs for test-source-id's app drain with url dropping://my-drain:8080/path
  2025/02/24 15:13:59 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  [FAILED] in [It] - /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:102 @ 02/24/25 15:13:59.6
  << Timeline

  [FAILED] Timed out after 0.189s.
  Expected
      <int>: 0
  to equal
      <int>: 10
  In [It] at: /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:102 @ 02/24/25 15:13:59.6

  Full Stack Trace
    code.cloudfoundry.org/loggregator-agent-release/src/pkg/egress/syslog_test.init.func3.3()
    	/tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:102 +0x668
------------------------------
••••

Summarizing 2 Failures:
  [FAIL] HTTPS_batch [It] testing simple appending of one log
  /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:69
  [FAIL] HTTPS_batch [It] test batch dispatching with all logs in a given timeframe
  /tmp/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:102

Ran 84 of 84 Specs in 3.763 seconds
FAIL! -- 82 Passed | 2 Failed | 0 Pending | 0 Skipped
```
These two tests fail randomly.

Please squash your commits  as well.

chombium reviewed (dismissed):
LGTM! @nicklas-dohrn thanks for your contribution and the cooperation.

ctlong commented:
Pipeline unit tests failed again:

```
• [FAILED] [0.216 seconds]
HTTPS_batch [It] test batch dispatching with all logs in a given timeframe
/tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:98

  Timeline >>
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  2025/03/05 20:03:18 Dropped 10000 dropping logs for app-id's app drain with url dropping:
  [FAILED] in [It] - /tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:104 @ 03/05/25 20:03:18.925
  2025/03/05 20:03:18 Dropped 10000 dropping logs for aggregate drain with url dropping:
  << Timeline

  [FAILED] Timed out after 0.185s.
  Expected
      <int>: 0
  to equal
      <int>: 5
  In [It] at: /tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:104 @ 03/05/25 20:03:18.925

  Full Stack Trace
    code.cloudfoundry.org/loggregator-agent-release/src/pkg/egress/syslog_test.init.func3.4()
    	/tmp/build/f541ec31/loggregator-agent-release/src/pkg/egress/syslog/https_batch_test.go:104 +0x61a
```

ctlong commented:
I'm retriggering it to see if it succeeds on a second run

ctlong commented on src/pkg/egress/syslog/https_batch_test.go:
```diff
@@ -83,37 +96,48 @@ var _ = Describe("HTTPS_batch", func() {
-		time.Sleep(100 * time.Millisecond)
-		Expect(drain.getMessagesSize()).Should(Equal(10))
+		Expect(drain.getMessagesSize()).To(Equal(0))
+		Eventually(drain.getMessagesSize, 180*time.Millisecond).Should(Equal(5))
```
Can you please make this timeout a big longer to account for some processing time.

nicklas-dohrn replied:
I increased the timeouts for all tests by at least 2x.

ctlong commented on src/pkg/egress/syslog/https_batch_test.go:
```diff
@@ -83,37 +96,48 @@ var _ = Describe("HTTPS_batch", func() {
+		}
+		// This is a less flaky approach to test for batch size triggered sends:
+		// with only Time based sends, it either takes
+		Eventually(drain.getMessagesSize, 180*time.Millisecond).Should(Equal(15))
```
Can you please make this timeout a big longer to account for some processing time.

nicklas-dohrn commented:
I increased the timeout on all tests by 2x at least
And made it consistent for all tests.
This should now proceed.

juergen-walter commented:
All tests have passed @ctlong @chombium, I think it is ready to be merged.

chombium commented:
Hi @carson,

As we cannot always reproduce the test failures and we (everyone except VMware employees) don't see what's happening in the CI/CD release pipeline, I've done few more tests in VMs with low resources running Ubuntu Server 22.04
I tested with two VMs with:
- 4 CPUs and 8 gb RAM and 
- 3 CPUs and 4 gb RAM (I wanted to test with 2 CPUs, but the VMs didn't start)

As I couldn't see the tests failing, can you please check i the tests with the current changes in the PR run stable in the CI/CD release pipeline? If not can you please tell us what is the exact specifications of the VM running the validation, so that we I try to reproduce the problems?

chombium commented:
Thanks for the help @ctlong
